### ISSUE TITLE
Questions on ATEPC (test) metrics | Evaluation training procedure

### ISSUE THREAD

**Issue by KarstenLasse** [yangheng95/PyABSA#228]
Hi there @yangheng95 , 
I have a few questions about evaluation of ATEPC models which are also relevant for other users. It would be great if you could provide me some clarity on the following (interrelated) things

1. _Most importantly_, after doing a training procedure with `Trainer()`, a Metric Visualizer (MV) table is generated with max-test acc. APC, max-test F1-APC and max test-F1-ATE. Are these metrics indeed on (holdout) Test data? 
[EXAMPLE_1]
`INFO:  ------------------------------------- Metric Visualizer ------------------------------------- 
╒══════════════╤══════════════════════════╤════════════════════════════╤═════════════════════════════════════════════════════════════════════════╕
│ Metric       │ Model & Dataset          │ Values (First 10 values)   │ Summary                                                                 │
╞══════════════╪══════════════════════════╪════════════════════════════╪═════════════════════════════════════════════════════════════════════════╡
│ Test-APC-Acc │ LCF_ATEPC-custom_dataset │ [75.91]                    │ ['Avg:75.91, Median: 75.91, IQR: 0.0, STD:0.0, Max: 75.91, Min: 75.91'] │
├──────────────┼──────────────────────────┼────────────────────────────┼─────────────────────────────────────────────────────────────────────────┤
│ Test-APC-F1  │ LCF_ATEPC-custom_dataset │ [53.51]                    │ ['Avg:53.51, Median: 53.51, IQR: 0.0, STD:0.0, Max: 53.51, Min: 53.51'] │
├──────────────┼──────────────────────────┼────────────────────────────┼─────────────────────────────────────────────────────────────────────────┤
│ Test-ATE-F1  │ LCF_ATEPC-custom_dataset │ [96.06]                    │ ['Avg:96.06, Median: 96.06, IQR: 0.0, STD:0.0, Max: 96.06, Min: 96.06'] │
╘══════════════╧══════════════════════════╧════════════════════════════╧═════════════════════════════════════════════════════════════════════════╛` (By the way, Median and IQR do not seem to work properly yet, maybe you're still working on this.
(See point below for confusion).

2. [Potential bug] After the trainingprocedure also an INFO list is generated with therein in my case. [EXAMPLE_2] `INFO: max_test_metrics:{'max_apc_test_acc': 82.24, 'max_apc_test_f1': 54.69, 'max_ate_test_f1': 96.44}`. However these metrics are not the same as the metrics in the MV table. Rather, the metrics in the INFO output are the same as the highest metrics outputted during training (generated per epoch); in my case [EXAMPLE_3] `100%|██████████| 260/260 [02:50<00:00,  1.52it/s, Epoch:9 | loss_apc:0.1587 | loss_ate:0.0225 | APC_ACC: 75.87(**max:82.24**) | APC_F1: 50.02(**max:54.69**) | ATE_F1: 95.95(**max:96.44**)]` Why do the INFO test metrics [EXAMPLE 2] differ from the max_test_metrics [EXAMPLE 1] of the MV table (which I think are actually on test, since they're lower)? And why do the INFO test metrics copy the metrics generated during training (which should be on the _validation_ set and aren't _test_ metrics according to best practices)?
(Please see point below)

3. Are the metrics generated during training procedure (so those metrics per epoch) based on the validation set or training set? (What is extra confusing is that the metrics in the INFO output [EXAMPLE 2] refers to these max. metrics from the training procedure [EXAMPLE 3]. I take it that the test data is **not** used as evaluation during the training procedure otherwise the test data would be contaminated and not held-out anymore.

4. Is it possible to steer the model during training towards improving a certain metric if using patience (or at all), e.g. towards improving the APC_F1 score. If possible that would be a cool feature to add, and improve usability. (I thought it was possible on Transformer models, but I might be wrong).

5. [Suggestion, not urgent] (Unrelated to above) Many files that the package generates, have . (dots) in their names. Is there a specific reason for that, because it is not best practice to use dots filenames in python language (as dots are used for methods/attributes), and can result in problems (especially for Windows systems). Why not use underscores _ ? Just a suggestion :)

Rest me to say that the ATEPC infrastructure is really great. Many research is focused on Aspect-Target Sentiment Classification only. However in reality being able to find both aspects as sentiment accompanied with these aspects is of major importance. So thank you.

Clarification on the matter would be hugely appreciated. Thanks in advance!

Kind regards, Karsten

**Comment by yangheng95** [comment]
Hi, 
first of all, the IQR and Median only works on enough trials. The max_text*** is original used to log the best performance of each metric. However, final experimental results in MV is the global performance. If you dont use a validation set, then both two results are obtained from testing set. If you have a validation set, it is expected to evaluate on the validation set in INFO. And in this case, the results are derived from testing set. In summary, the results in MV is always evaluated on testing set.

**Comment by KarstenLasse** [comment]
Allright all good, great. And since I use a validation set, are the epoch F1-scores during training then also on based the validation data? And can I find in somewhere in the code how to set up trials? Haven't run into it yet.

**Comment by yangheng95** [comment]
That's right, you can set multiple random seed to run different trials, each random seed represents for a new trial.

**Comment by yangheng95** [comment]
evaluate on valdiation set: https://github.com/yangheng95/PyABSA/blob/964d7862da13ef8cc38cb56fe0e65086b343a9cd/pyabsa/core/atepc/training/atepc_trainer.py#L288

Please check the training log  to see if the validation set is detected correctly.

**Comment by KarstenLasse** [comment]
Thanks!
